realm-transformer/src/main/kotlin/io/realm/transformer/RealmTransformer.kt

```diff
@@ -0,0 +1,155 @@
+                QualifiedContent.Scope.TESTED_CODE
+        )
+    }
+
+    /**
+     * Implements the transform algorithm. The heaviest part of the transform is loading the
+     * {@code CtClass} from JavaAssist, so this should be avoided as much as possible.
+     *
+     * This is also the reason that there are significant changes between a full build and a
+     * incremental build. In a full build, we can use text matching to go from a proxy class
+     * to the model class. Something we cannot do when building incrementally. In that case
+     * we have to deduce all information from the class at hand.
+     *
+     * @param context
+     * @param inputs
+     * @param referencedInputs
+     * @param outputProvider
+     * @param isIncremental
+     * @throws IOException
+     * @throws TransformException
+     * @throws InterruptedException
+     */
+    override fun transform(context: Context?, inputs: MutableCollection<TransformInput>?,
+                           referencedInputs: Collection<TransformInput>?,
+                           outputProvider: TransformOutputProvider?, isIncremental: Boolean) {
+
+        val tic = System.currentTimeMillis()
+
+        val build: TransformBuildTemplate = if (isIncremental) IncrementalBuild(project, outputProvider!!, this)
+        else FullBuild(project, outputProvider!!, this)
+
+        build.prepareOutputClasses(inputs!!);
+        if (build.hasNoOutput()) {
+            // Abort transform as quickly as possible if no files where found for processing.
+            exitTransform(emptySet(), emptyList(), tic)
+            return
+        }
+        build.prepareReferencedClasses(referencedInputs!!);
+        build.markMediatorsAsTransformed()
+        build.transformModelClasses();
+        build.transformDirectAccessToModelFields();
+        build.copyResourceFiles();
+        exitTransform(inputs, build.getOutputModelClasses(), tic)
+    }
+
+    private fun exitTransform(inputs: Collection<TransformInput>, outputModelClasses: Collection<CtClass>, startTime: Long) {
+        val endTime = System.currentTimeMillis()
+        logger.debug("Realm Transform time: ${endTime-startTime} milliseconds")
+        this.sendAnalytics(inputs, outputModelClasses)
+    }
+
+    /**
+     * Sends the analytics
+     *
+     * @param inputs the inputs provided by the Transform API
+     * @param inputModelClasses a list of ctClasses describing the Realm models
+     */
+    private fun sendAnalytics(inputs: Collection<TransformInput>, outputModelClasses: Collection<CtClass>) {
+        val disableAnalytics: Boolean = "true".equals(System.getenv()["REALM_DISABLE_ANALYTICS"])
+        if (inputs.isEmpty() || disableAnalytics) {
+            // Don't send analytics for incremental builds or if they have ben explicitly disabled.
+            return
+        }
+
+        var containsKotlin = false
+
+        outer@
+        for(input: TransformInput in inputs) {
```

Comment:
Not sure what you mean? The transform is single-threaded because you cannot exit it until everything inside is done anyway. Otherwise you would risk the next transform in line

Reply:
I mean the double for loop we're doing to figure out if it's a Kotlin build, can it be moved inside the worker thread that sends the analytics, since it's only used there

Reply:
Ah yes, that is a good idea 👍